lock-api/src/main/java/com/palantir/lock/client/ProfilingTimelockService.java

```diff
@@ -0,0 +1,228 @@
+        return runTaskTimed("getFreshTimestamps", () -> delegate.getFreshTimestamps(numTimestampsRequested));
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp() {
+        // N.B. Immutable timestamp lock is not exclusive, so it should be fast.
+        return runTaskTimed("lockImmutableTimestamp", delegate::lockImmutableTimestamp);
+    }
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startIdentifiedAtlasDbTransaction() {
+        return runTaskTimed("startIdentifiedAtlasDbTransaction", delegate::startIdentifiedAtlasDbTransaction);
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return runTaskTimed("getImmutableTimestamp", delegate::getImmutableTimestamp);
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        // Don't profile this, as it may be skewed by user contention on locks.
+        tryFlushLogs();
+        return delegate.lock(request);
+    }
+
+    @Override
+    public WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        // Don't profile this, as it may be skewed by user contention on locks.
+        tryFlushLogs();
+        return delegate.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockToken> refreshLockLeases(Set<LockToken> tokens) {
+        return runTaskTimed("refreshLockLeases", () -> delegate.refreshLockLeases(tokens));
+    }
+
+    @Override
+    public Set<LockToken> unlock(Set<LockToken> tokens) {
+        return runTaskTimed("unlock", () -> delegate.unlock(tokens));
+    }
+
+    @Override
+    public long currentTimeMillis() {
+        return runTaskTimed("currentTimeMillis", delegate::currentTimeMillis);
+    }
+
+    private <T> T runTaskTimed(String actionName, Supplier<T> action) {
+        Stopwatch stopwatch = stopwatchSupplier.get();
+        try {
+            T result = action.get();
+            trackSuccessfulAction(actionName, stopwatch);
+            return result;
+        } catch (RuntimeException e) {
+            trackFailedAction(actionName, stopwatch, e);
+            throw e;
+        }
+    }
+
+    private void trackSuccessfulAction(String actionName, Stopwatch stopwatch) {
+        trackActionAndMaybeLog(actionName, stopwatch, Optional.empty());
+    }
+
+    private void trackFailedAction(String actionName, Stopwatch stopwatch, RuntimeException exception) {
+        trackActionAndMaybeLog(actionName, stopwatch, Optional.of(exception));
+    }
+
+    private void trackActionAndMaybeLog(String actionName, Stopwatch stopwatch, Optional<Exception> failure) {
+        stopwatch.stop();
+        accumulateSlowOperationTracking(actionName, stopwatch, failure);
+        tryFlushLogs();
+    }
+
+    private void accumulateSlowOperationTracking(String actionName, Stopwatch stopwatch, Optional<Exception> failure) {
+        if (stopwatchDescribesSlowOperation(stopwatch)) {
+            slowestOperation.accumulateAndGet(
```

Comment:
should have mentioned this earlier, do we want to add a meter as well? or something that we can get a dashboard of more easily?

Reply:
I think in general you have `TimeLockService.<op>` for individual operations (though the granularity there is not as fine, and you also can't check failures). I'd leave this for a separate PR if we wanted it?